src/hotspot/share/gc/parallel/psParallelCompact.cpp

```diff
@@ -1449,6 +1449,18 @@ static void split_regions_for_worker(size_t start, size_t end,
                 + (worker_id < remainder ? 1 : 0);
 }
 
+static bool safe_to_read_header(size_t words) {
+  precond(words > 0);
+
+  const bool safe = words >= (size_t)oopDesc::header_size();
```

Comment:
I don't know what this means.  Does header_size() include the markWord and klass pointer? ie 2 words for !UseCompactObjectHeaders? 
Is it safe if the number of words includes a klass pointer?
Can you enclose EnableValhalla in this method rather than asking in these callers?

Reply:
If you move EnableValhalla to inside safe_to_read_header(), then we'll only have to fix it once when it's only set when Arguments::is_valhalla_enabled() is set.  Right now EnablePreview is true by default.
EnableValhalla is a property that determines the answer to safe_to_read_header(), so I think it should go there.

Can you add a comment to safe_to_read_header() - the second sentence to the first paragraph above?